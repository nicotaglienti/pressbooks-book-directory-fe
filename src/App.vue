<template>
  <v-app class="mx-auto">
    <header-bar></header-bar>
    <v-container class="v-application__main_content" fluid>
      <ais-instant-search
        :search-client="$store.state.SClient.searchClient"
        :index-name="$store.state.SClient.indexName"
      >
        <ais-configure :hits-per-page.camel="$store.state.SClient.searchParameters.hitsPerPage"></ais-configure>
        <welcome-header></welcome-header>
        <current-filters></current-filters>
        <v-row no-gutters>
          <v-col cols="6" md="3">
            <filters></filters>
          </v-col>
          <v-col cols="12" md="9">
            <div v-if="$store.state.config.canFilter">
              <books></books>
              <pagination></pagination>
            </div>
          </v-col>
        </v-row>
      </ais-instant-search>
    </v-container>
  </v-app>
</template>

<script>
  import "instantsearch.css/themes/algolia-min.css";
  import Filters from "./components/filters/Filters";
  import Pagination from "./components/commons/Pagination";
  import HeaderBar from "./components/commons/HeaderBar";
  import CurrentFilters from "./components/filters/CurrentFilters";
  import Books from "./components/Books";
  import WelcomeHeader from "./components/WelcomeHeader";

  export default {
    components: {
      WelcomeHeader,
      CurrentFilters,
      Filters,
      Pagination,
      HeaderBar,
      Books
    },
    mounted() {
      let index = this.$store.state.SClient.searchClient.initIndex(this.$store.state.SClient.indexName);
      this.$store.dispatch('getStats', index);
    }
  };
</script>
<style lang="sass">
  @import 'styles/styles'
</style>