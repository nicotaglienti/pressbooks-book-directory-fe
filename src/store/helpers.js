function setFilters(oldFilters, allowedFilters) {
    let fs = [], ns = [], strQuery = '', f = {};
    let query = [], queryExcluded = [], numerics = {}, q =[];
    let qtyAttr = Object.keys(oldFilters).length;
    for (let attribute in oldFilters) {
        q = [];
        for (let i =0; i < oldFilters[attribute].length; i++) {
            f = oldFilters[attribute][i];
            strQuery = ':';
            if (f.exclude) {
                if (allowedFilters[attribute].type !== 'boolean') {
                    strQuery = ':-';
                }
                queryExcluded.push(f.attribute + strQuery + f.value);
            } else {
                if (f.operator !== undefined) {
                    if (numerics[f.attribute] === undefined) {
                        numerics[f.attribute] = [];
                    }
                    numerics[f.attribute].push(f);
                } else {
                    q.push(f.attribute + strQuery + f.value);
                }
            }
        }
        if (q.length > 0) {
            query.push(q);
        }
    }
    if (queryExcluded.length > 0 && query.length > 0) {
        if (qtyAttr === 1) {
            fs.push(query);
        } else {
            fs = query;
        }
        fs = fs.concat(queryExcluded);
    }
    if (query.length > 0 && queryExcluded.length === 0) {
        fs = fs.concat(query);
    }
    if (query.length === 0 && queryExcluded.length > 0) {
        fs = fs.concat(queryExcluded);
    }
    for (let attr in numerics) {
        if (numerics[attr].length > 1) {
            let v1 = numerics[attr][0].value;
            let v2 = numerics[attr][1].value;
            if (attr === 'storageSize') {
                v1 = numerics[attr][0].value * 1024 * 1024;
                v2 = numerics[attr][1].value * 1024 * 1024;
            }
            ns.push(attr + ':' + v1 + ' TO ' + v2);
        } else {
            let v3 = numerics[attr][0].value;
            if (attr === 'storageSize') {
                v3 = numerics[attr][0].value * 1024*1024;
            }
            strQuery = numerics[attr][0].operator;
            ns.push(attr + strQuery + v3);
        }
    }
    return [fs, ns.join(' AND ')];
}

export default {
    functions: {
        setFilters
    }
};