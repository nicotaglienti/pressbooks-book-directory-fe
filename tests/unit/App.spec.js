import Vue from "vue";
import Vuetify from "vuetify";
import InstantSearch from "vue-instantsearch";
import App from "@/App";
import { store } from "@/store/index";

Vue.use(Vuetify);
Vue.use(InstantSearch);

import {
    mount
} from "@vue/test-utils";

describe("Check main structure for App.vue", () => {
    let vuetify;

    vuetify = new Vuetify({
        theme: { dark: true },
        rtl: false
    });

    const wrapper = mount(App, {
        Vue,
        vuetify,
        store
    });

    it("should have AisInstantSearch component", () => {
        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.findComponent({ name: "AisInstantSearch" }).exists()).toBe(true);
    });

    it("should have WelcomeHeader component", () => {
        const welcomeHeader = wrapper.findComponent({ name: "WelcomeHeader" });
        expect(welcomeHeader.exists()).toBe(true);
    });

    it("should have CurrentFilters component", () => {
        const currentFilters = wrapper.findComponent({ name: "CurrentFilters" });
        expect(currentFilters.exists()).toBe(true);
    });

    it("should have Filters component", () => {
        const Filters = wrapper.findComponent({ name: "Filters" });
        expect(Filters.exists()).toBe(true);
    });

    it("should have Books component", () => {
        const Books = wrapper.findComponent({ name: "Books" });
        expect(Books.exists()).toBe(true);
    });

    it("should have Pagination component", () => {
        const Pagination = wrapper.findComponent({ name: "Pagination" });
        expect(Pagination.exists()).toBe(true);
    });
});