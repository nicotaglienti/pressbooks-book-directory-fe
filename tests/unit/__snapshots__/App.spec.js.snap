// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check main structure for App.vue should have AisInstantSearch component 1`] = `
<div data-app="true" class="v-application v-application--is-ltr theme--dark" id="app">
  <div class="v-application--wrap">
    <header class="toolbar v-sheet v-sheet--tile theme--dark v-toolbar" style="height: 64px;">
      <div class="v-toolbar__content" style="height: 64px;"><a href="https://pressbooks.com" aria-label="Visit Pressbooks.com" target="_blank">
          <div class="v-responsive v-image logo">
            <transition-stub name="fade-transition" mode="in-out">
              <div class="v-image__image v-image__image--preload v-image__image--cover" style="background-position: center center;"></div>
            </transition-stub>
            <div class="v-responsive__content"></div>
          </div>
        </a> <a href="https://twitter.com/intent/follow?screen_name=pressbooks" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--dark v-size--default" aria-label="Visit Pressbooks' Twitter account"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-twitter theme--dark"></i></span></a></div>
    </header>
    <main role="main">
      <div class="container" id="v-application__main_content">
        <div class="ais-InstantSearch">
          <!----> <button type="button" class="float-right v-btn v-btn--contained theme--dark v-size--default" aria-label="Take the tour"><span class="v-btn__content">
          Take the tour
        </span></button>
          <div class="container welcome-header">
            <h1>Welcome to Pressbooks Directory</h1>
            <p>
              This directory provides an index of <strong>0</strong> books published across
              <strong>0</strong> Pressbooks networks. Search and filter books by keyword,
              subject matter, license, and more. See
              <a href="https://networkmanagerguide.pressbooks.com/chapter/how-to-use-the-pressbooks-directory/">our guide for
                more detailed instructions</a> on using the Pressbooks Directory to find books of interest.
            </p>
            <!---->
            <!---->
            <!---->
            <form novalidate="novalidate" class="v-form">
              <div class="row justify-space-between">
                <div class="col-md-2 col-12">
                  <h2 class="ais-SearchBox__text">
                    Find a book:
                  </h2>
                </div>
                <div class="col-md-9 col-12">
                  <div class="v-input theme--dark v-text-field">
                    <div class="v-input__control">
                      <div class="v-input__slot">
                        <div class="v-text-field__slot"><label for="search-book" class="v-label theme--dark" style="left: 0px; position: absolute;">Search all book metadata</label><input id="search-book" type="search"></div>
                        <div class="v-input__append-inner"><i aria-hidden="true" class="v-icon notranslate mdi mdi-magnify theme--dark"></i></div>
                      </div>
                      <div class="v-text-field__details">
                        <div class="v-messages theme--dark">
                          <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div align="center" class="col-md-1 col-12"><button type="submit" class="v-btn v-btn--contained theme--dark v-size--default" id="search-button"><span class="v-btn__content">
        Search
      </span></button></div>
              </div>
            </form>
          </div>
          <div id="current-filters">
            <div class="row filters">
              <div class="filters__head filters__head--red col-md-9 col-12">
                Active Filters:
                <button type="reset" disabled="disabled" class="ais-ClearRefinements-button ais-ClearRefinements-button--disabled">
                  Clear refinements
                </button>
                <!---->
              </div>
              <div class="col-md-3 col-12">
                <!---->
                <!---->
                <!---->
              </div>
            </div>
            <div class="text-center">
              <!---->
            </div>
          </div>
          <div class="row">
            <div id="filters" class="col-md-4 col-lg-3 col-12">
              <div class="search-panel__filters">
                <div role="list" class="v-list v-sheet v-sheet--tile theme--dark v-list--dense">
                  <!---->
                  <div class="v-list-group v-list-group--sub-group" id="filter-license_code">
                    <div tabindex="0" aria-expanded="false" role="button" class="v-list-group__header v-list-item v-list-item--link theme--dark">
                      <div class="v-list-item__icon v-list-group__header__prepend-icon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--dark"></i></div>
                      <div class="v-list-item__title">LICENSE</div>
                    </div>
                    <transition-stub name="expand-transition">
                      <!---->
                    </transition-stub>
                  </div>
                  <div class="v-list-group v-list-group--sub-group" id="filter-about">
                    <div tabindex="0" aria-expanded="false" role="button" class="v-list-group__header v-list-item v-list-item--link theme--dark">
                      <div class="v-list-item__icon v-list-group__header__prepend-icon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--dark"></i></div>
                      <div class="v-list-item__title">SUBJECT</div>
                    </div>
                    <transition-stub name="expand-transition">
                      <!---->
                    </transition-stub>
                  </div>
                  <div class="v-list-group v-list-group--sub-group" id="filter-networkName">
                    <div tabindex="0" aria-expanded="false" role="button" class="v-list-group__header v-list-item v-list-item--link theme--dark">
                      <div class="v-list-item__icon v-list-group__header__prepend-icon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--dark"></i></div>
                      <div class="v-list-item__title">NETWORK</div>
                    </div>
                    <transition-stub name="expand-transition">
                      <!---->
                    </transition-stub>
                  </div>
                  <!---->
                  <div class="v-list-group v-list-group--sub-group" id="filter-lastUpdated">
                    <div tabindex="0" aria-expanded="false" role="button" class="v-list-group__header v-list-item v-list-item--link theme--dark">
                      <div class="v-list-item__icon v-list-group__header__prepend-icon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--dark"></i></div>
                      <div class="v-list-item__title">LAST UPDATED</div>
                    </div>
                    <transition-stub name="expand-transition">
                      <!---->
                    </transition-stub>
                  </div>
                  <div class="v-list-group v-list-group--sub-group" id="filter-languageName">
                    <div tabindex="0" aria-expanded="false" role="button" class="v-list-group__header v-list-item v-list-item--link theme--dark">
                      <div class="v-list-item__icon v-list-group__header__prepend-icon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--dark"></i></div>
                      <div class="v-list-item__title">LANGUAGE</div>
                    </div>
                    <transition-stub name="expand-transition">
                      <!---->
                    </transition-stub>
                  </div>
                  <div class="v-list-group v-list-group--sub-group" id="filter-publisher_name">
                    <div tabindex="0" aria-expanded="false" role="button" class="v-list-group__header v-list-item v-list-item--link theme--dark">
                      <div class="v-list-item__icon v-list-group__header__prepend-icon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--dark"></i></div>
                      <div class="v-list-item__title">PUBLISHER</div>
                    </div>
                    <transition-stub name="expand-transition">
                      <!---->
                    </transition-stub>
                  </div>
                  <div class="v-list-group v-list-group--sub-group" id="filter-wordCount">
                    <div tabindex="0" aria-expanded="false" role="button" class="v-list-group__header v-list-item v-list-item--link theme--dark">
                      <div class="v-list-item__icon v-list-group__header__prepend-icon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--dark"></i></div>
                      <div class="v-list-item__title">WORD COUNT</div>
                    </div>
                    <transition-stub name="expand-transition">
                      <!---->
                    </transition-stub>
                  </div>
                  <div class="v-list-group v-list-group--sub-group" id="filter-h5pActivities">
                    <div tabindex="0" aria-expanded="false" role="button" class="v-list-group__header v-list-item v-list-item--link theme--dark">
                      <div class="v-list-item__icon v-list-group__header__prepend-icon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--dark"></i></div>
                      <div class="v-list-item__title">H5P ACTIVITIES</div>
                    </div>
                    <transition-stub name="expand-transition">
                      <!---->
                    </transition-stub>
                  </div>
                  <div class="v-list-group v-list-group--sub-group" id="filter-basedOn">
                    <div tabindex="0" aria-expanded="false" role="button" class="v-list-group__header v-list-item v-list-item--link theme--dark">
                      <div class="v-list-item__icon v-list-group__header__prepend-icon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--dark"></i></div>
                      <div class="v-list-item__title">BASED ON</div>
                    </div>
                    <transition-stub name="expand-transition">
                      <!---->
                    </transition-stub>
                  </div>
                  <div class="v-list-group v-list-group--sub-group" id="filter-storageSize">
                    <div tabindex="0" aria-expanded="false" role="button" class="v-list-group__header v-list-item v-list-item--link theme--dark">
                      <div class="v-list-item__icon v-list-group__header__prepend-icon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--dark"></i></div>
                      <div class="v-list-item__title">STORAGE SIZE</div>
                    </div>
                    <transition-stub name="expand-transition">
                      <!---->
                    </transition-stub>
                  </div>
                </div>
              </div>
            </div>
            <div id="book-list" class="col-md-8 col-lg-9 col-12">
              <div>
                <!---->
              </div>
            </div>
          </div>
          <div class="text-center">
            <!---->
          </div>
        </div> <a href="https://www.algolia.com/" aria-label="Search by Algolia" target="_blank" class="filters__stats-algolia-logo"></a>
      </div>
      <!---->
    </main>
  </div>
</div>
`;
